### PR TITLE
Store cluster configuration in microcluster

### PR DESCRIPTION
**Pull request by neoaggelos**
### Summary

This is a huge PR that cleans up the codebase and aims to unblock things going forward.

### Changes 

By order of importance:

- We set `ClusterConfig` during bootstrap, and consume it when joining control plane nodes
- We update the `ClusterConfig` when enabling DNS (to set the new kubelet configs), and consume it when enabling `network`
- `Snap interface` is now an abstraction over all filesystem path details. Whenever this info is needed, we pass the `snap`  instance. Currently the interface also as a service manager, we might want to extract it at a later time. All filesystem operations are moved outside of the `Snap` interface (the PR currently leaves things a bit flat atm, we can add more structure later.
- Add individual functions to configure each component (moved to `pkg/k8sd/setup`) (e.g. containerd, k8s-dqlite, kubernetes services) etc. These apply configurations based on the active cluster configuration. Also adds unit tests for containerd (tests for other services can be added at a later time)
- Rework certificate handling (see `k8sd/pki` package). Adds ground-work for using external certificates (if provided), or generate self-signed ones by default.
- Move service arguments to `$SNAP_COMMON`, simplifying some config scenarios.
- Add proper handling for `k8s-dqlite` datastore when joining control plane nodes.

### Bug fixes

A number of bugs are fixed by the work on this PR:

- Ensure worker nodes get a proper `kubelet` certificate, which fixes apiserver -> kubelet communication (e.g. `kubectl logs` commands failing for worker nodes because of invalid certificate SANs).
- Enforce the hostname and node IP addresses for kubelet, kube-proxy and k8s-dqlite. Previously, these were left unset (using node defaults), which could cause unforeseen problems on clusters. Some examples:
  - k8s-dqlite binding on a different interface
  - kubelet node IP changing due to cilium
  - kubelet node IP changing after reboot, causing cascading failures
  - hostname changes breaking kubelet and kube-proxy, and registering a second node instead.

### CODE REVIEW

File: src/k8s/pkg/component/component.go [hunk 1]
```diff
@@ -16,18 +16,9 @@ import (
...
-	ReleaseName     string `mapstructure:"release"`
-	Chart           string `mapstructure:"chart"`
-	Namespace       string `mapstructure:"namespace"`
-}
-
 // helmClient implements the ComponentManager interface
 type helmClient struct {
-	config      map[string]componentDefinition
-	snap        snap.Snap
+	components  map[string]types.Component
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/component/component.go:21]: this comes from the internal `types` package, so it should be OK. Does not affect the API

File: src/k8s/pkg/component/component_test.go [hunk 2]
```diff
@@ -192,9 +194,9 @@ func TestListComponentsWithReleases(t *testing.T) {
 	components, err := mockHelmClient.List()
 
 	g.Expect(err).To(BeNil())
-	g.Expect(components).To(HaveLen(3))
-
-	g.Expect(components[0]).To(Equal(Component{Name: "one", Status: true}))
-	g.Expect(components[2]).To(Equal(Component{Name: "two", Status: true}))
-	g.Expect(components[1]).To(Equal(Component{Name: "three", Status: true}))
+	g.Expect(components).To(Equal([]Component{
+		{Name: "one", Status: true},
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/component/component_test.go:198]: indeed, simply preferred not to change this. also, API representation is OK to be a string, the implementation can iterate and we may need for status going forward (e.g. "NeedsUpdate")

File: src/k8s/pkg/k8s/client/cluster_node.go [hunk 3]
```diff
@@ -78,5 +77,6 @@ func (c *Client) CleanupNode(ctx context.Context, nodeName string) {
 	// joining another cluster.
 	c.ResetNode(ctx, nodeName, true)
 
-	snap.StopControlPlaneServices(ctx, snap.NewDefaultSnap())
+	// TODO(neoaggelos): reenable after we know how to pass a snap here
+	// snap.StopControlPlaneServices(ctx, snap.NewDefaultSnap())
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/k8s/client/cluster_node.go:81]: good point

File: src/k8s/pkg/k8s/client/cluster.go [hunk 4]
```diff
@@ -46,6 +46,7 @@ func (c *Client) Bootstrap(ctx context.Context, bootstrapConfig apiv1.BootstrapC
 	}
 	if err := c.m.NewCluster(hostname, addrPort, config, time.Second*30); err != nil {
 		// TODO(neoaggelos): print message that bootstrap failed, and that we are cleaning up
+		fmt.Println("Cleaning up, error was", err)
```
Comment (bschimke95) [inline src/k8s/pkg/k8s/client/cluster.go:49]: nit: `fmt.Println("Cleaning up, error was: ", err)` makes it slightly more readable.

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 5]
```diff
@@ -137,93 +143,120 @@ func onBootstrapWorkerNode(state *state.State, encodedToken string) error {
...
+	if err := setup.KubeScheduler(snap); err != nil {
+		return fmt.Errorf("failed to configure kube-scheduler: %w", err)
 	}
-	clusterConfig.Certificates.K8sDqliteCert = string(k8sDqliteCertPair.CertPem)
-	clusterConfig.Certificates.K8sDqliteKey = string(k8sDqliteCertPair.KeyPem)
-
-	caPair, err := cert.LoadCertKeyPair(path.Join(cert.KubePkiPath, "ca.key"), path.Join(cert.KubePkiPath, "ca.crt"))
-	if err != nil {
-		return fmt.Errorf("failed to load k8s-dqlite cert-key pair: %w", err)
+	if err := setup.KubeAPIServer(snap, cfg.Network.ServiceCIDR, s.Address().Path("1.0", "kubernetes", "auth", "webhook").String(), true, cfg.APIServer.Datastore, cfg.APIServer.AuthorizationMode); err != nil {
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/app/hooks_bootstrap.go:231]: my preference as well, but `Path("1.0/kubernetes/auth/webhook")`, escapes the `/` characters...

File: src/k8s/pkg/k8sd/pki/control_plane.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+		allowSelfSignedCA: allowSelfSignedCA,
+		hostname:          hostname,
+		years:             years,
+		ipSANs:            ipSANs,
+		dnsSANs:           dnsSANs,
+	}
+}
+
+// CompleteCertificates generates missing or unset certificates. If only a certificate is set and not a key, we assume that the cluster is using managed certificates.
+func (c *ControlPlanePKI) CompleteCertificates() error {
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/pki/control_plane.go:50]: intentionally left flat, such that we can return to it and add structure at a later time as needed.

File: src/k8s/pkg/k8sd/app/hooks_join.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+		address := fmt.Sprintf("%s:%d", nodeIP.String(), cfg.K8sDqlite.Port)
+		if err := setup.K8sDqlite(snap, address, cluster); err != nil {
+			return fmt.Errorf("failed to configure k8s-dqlite with address=%s cluster=%v: %w", address, cluster, err)
+		}
+	default:
+		return fmt.Errorf("unknown datastore %q, must be k8s-dqlite", cfg.APIServer.Datastore)
+	}
+
+	// Configure services
+	if err := setup.Containerd(snap); err != nil {
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/app/hooks_join.go:106]: intentionally left flat, we can add structure at a later time as needed.

File: src/k8s/pkg/k8sd/app/hooks_join.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+	if err := certificates.CompleteCertificates(); err != nil {
+		return fmt.Errorf("failed to initialize cluster certificates: %w", err)
+	}
+	if err := setup.EnsureControlPlanePKI(snap, certificates); err != nil {
+		return fmt.Errorf("failed to write cluster certificates: %w", err)
+	}
+
+	// Generate kubeconfigs
+	for _, kubeconfig := range []struct {
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/app/hooks_join.go:61]: intentionally left flat, such that we can add more structure to this as needed.

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 9]
```diff
@@ -31,7 +34,7 @@ func onBootstrap(s *state.State, initConfig map[string]string) error {
 	return onBootstrapControlPlane(s, initConfig)
 }
 
-func onBootstrapWorkerNode(state *state.State, encodedToken string) error {
+func onBootstrapWorkerNode(s *state.State, encodedToken string) error {
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/app/hooks_bootstrap.go:37]: there's a lot of places around the code where we don't enforce this, which makes for some confusion, let's return to this at a later time.

File: src/k8s/pkg/k8sd/setup/certificates.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	if certificates.CAKey != "" {
+		toWrite[path.Join(snap.KubernetesPKIDir(), "ca.key")] = certificates.CAKey
+	}
+	if certificates.FrontProxyCAKey != "" {
+		toWrite[path.Join(snap.KubernetesPKIDir(), "front-proxy-ca.key")] = certificates.FrontProxyCACert
+	}
+
+	for fname, cert := range toWrite {
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/setup/certificates.go:36]: intentionally left flat at the moment, we can add structure later

**Review by berkayoz — COMMENTED** [review]
Amazing work, I've left some minor comments. I especially appreciate the unit tests setup.

File: src/k8s/cmd/k8sd/k8sd.go [hunk 11]
```diff
@@ -44,5 +46,5 @@ func init() {
 	rootCmd.PersistentFlags().BoolVarP(&rootCmdOpts.logDebug, "debug", "d", false, "Show all debug messages")
 	rootCmd.PersistentFlags().BoolVarP(&rootCmdOpts.logVerbose, "verbose", "v", true, "Show all information messages")
 	rootCmd.PersistentFlags().UintVar(&rootCmdOpts.port, "port", config.DefaultPort, "Port on which the REST API is exposed")
-	rootCmd.PersistentFlags().StringVar(&rootCmdOpts.stateDir, "state-dir", path.Join(os.Getenv("SNAP_COMMON"), "/var/lib/k8sd/state"), "Directory with the dqlite datastore")
+	rootCmd.PersistentFlags().StringVar(&rootCmdOpts.stateDir, "state-dir", "", "Directory with the dqlite datastore")
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/cmd/k8sd/k8sd.go:49]: yes, the default dynamically in code

File: src/k8s/pkg/k8sd/pki/control_plane_test.go [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+	c := pki.NewControlPlanePKI(pki.ControlPlanePKIOpts{
+		Hostname:          "h1",
+		Years:             10,
+		AllowSelfSignedCA: true,
+	})
+
+	g := NewWithT(t)
+
+	g.Expect(c.CompleteCertificates()).To(BeNil())
+	g.Expect(c.CompleteCertificates()).To(BeNil())
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/pki/control_plane_test.go:20]: yes, to ensures that if certificates are already generated, it's just a noop.

File: src/k8s/pkg/k8sd/pki/generate_test.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+	g.Expect(key).ToNot(BeEmpty())
+
+	t.Run("Load", func(t *testing.T) {
+		c, k, err := loadCertificate(cert, key)
+		g.Expect(err).To(BeNil())
+		g.Expect(c).ToNot(BeNil())
+		g.Expect(k).ToNot(BeNil())
+	})
+
+	t.Run("LoadKeyOnly", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/pki/generate_test.go:25]: right, updating

File: src/k8s/pkg/k8s/setup/k8s_dqlite.go [hunk 14]
```diff
@@ -85,28 +29,5 @@ func LeaveK8sDqliteCluster(ctx context.Context, snap snap.Snap, state *state.Sta
...
-	marshaled, err := yaml.Marshal(&init)
-	if err != nil {
-		return fmt.Errorf("failed to marshal cluster init data: %w", err)
-	}
-
-	if err := os.WriteFile(filepath.Join("/var/snap/k8s/common", cert.K8sDqlitePkiPath, "init.yaml"), []byte(marshaled), 0644); err != nil {
-		return fmt.Errorf("failed to write init.yaml to %s: %w", cert.K8sDqlitePkiPath, err)
-	}
-	return nil
+	return utils.RunCommand(ctx, "/snap/k8s/current/k8s/wrappers/commands/dqlite", "k8s", fmt.Sprintf(".remove %s", address))
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/k8s/setup/k8s_dqlite.go:32]: we should not shell out here at all, therefore adding a TODO to resolve this.

**Review by bschimke95 — APPROVED** [review]
LGTM

**Pull request merged**